Bintray Sunset

jonatan-ivanov opened the pull request:
Utilizing the `signing` and the `maven-publish` plugins to remove Bintray from the release process. 
Fixes gh-2455

shakuzen commented on build.gradle:
```diff
@@ -77,6 +79,37 @@ subprojects {
+            repositories {
+                maven {
+                    name = 'Milestone'
+                    url = 'https://repo.spring.io/milestone'
```
I don't know the difference between using the maven-publish plugin directly versus using the artifactory plugin to publish to this repository. The snapshot publish uses the artifactory plugin as far as I understand. Could we check and compare a snapshot vs candidate release with these changes?

jonatan-ivanov replied:
Since it is just a maven repo eventually, both will work, the artifactory plugin was a little more convenient since it was already set-up. Though, using the same process for all of our repos would make more sense to me too, also it will enable us to remove the artifactory plugin. I added the snapshot repo, the artifactory plugin is not used from now on and everything is published in the same way.

shakuzen commented on gradle/deploy.sh:
```diff
@@ -15,12 +15,10 @@ elif [ $CIRCLE_TAG ]; then
   *-rc\.*)
-    ./gradlew -Prelease.disableGitChecks=true -Prelease.useLastTag=true candidate bintrayUpload $SWITCHES -x release -x bintrayPublish
-    ./gradlew -Prelease.disableGitChecks=true -Prelease.useLastTag=true :micrometer-core:bintrayPublish $SWITCHES -x bintrayUpload
+    ./gradlew -Prelease.disableGitChecks=true -Prelease.useLastTag=true candidate publishAllPublicationsToMilestoneRepository $SWITCHES -x release -x artifactoryPublish -x bintrayUpload -x bintrayPublish
```
Nit: We could make `candidate` depend on the publish task in the build so the command here is simpler. I know it sounds a bit superfluous with all the skipped tasks here... We could follow-up getting rid of those with an update to the spring-release-plugin we're using.

jonatan-ivanov replied:
I was thinking on it too but eventually I did not do it since this way it is easier to test things locally and gives more information about what is happening. What do you think about cleaning this-up with the release plugin cleanup (that should also eliminate the excluded tasks)?

shakuzen replied:
Yes, I think it's fine to defer to when we change the release plugin. I think all of this is vetting the process before we bake it into a released version of the plugin.

shakuzen commented on gradle/deploy.sh:
```diff
@@ -15,12 +15,10 @@ elif [ $CIRCLE_TAG ]; then
   *-rc\.*)
-    ./gradlew -Prelease.disableGitChecks=true -Prelease.useLastTag=true candidate bintrayUpload $SWITCHES -x release -x bintrayPublish
-    ./gradlew -Prelease.disableGitChecks=true -Prelease.useLastTag=true :micrometer-core:bintrayPublish $SWITCHES -x bintrayUpload
+    ./gradlew -Prelease.disableGitChecks=true -Prelease.useLastTag=true candidate publishAllPublicationsToMilestoneRepository $SWITCHES -x release -x artifactoryPublish -x bintrayUpload -x bintrayPublish
```
publishAll is probably fine, but I wonder if it wouldn't be better to be more specific just in case there is more than one publication registered. Would we want to publish all of them? Can't we publish the nebula publication specifically since that's the one we expect to be there?

shakuzen reviewed (approved):
Looks good.
I think releasing the staged repository is still required for final releases and not covered yet. Perhaps we can consider the gradle nexus publish plugin for that. https://github.com/gradle-nexus/publish-plugin/

shakuzen commented on build.gradle:
```diff
@@ -77,6 +79,37 @@ subprojects {
+
+        signing {
+            useInMemoryPgpKeys(findProperty('SIGNING_KEY'), findProperty('SIGNING_PASSWORD'))
+            sign publishing.publications.nebula
```
Will our snapshot releases be signed too? I wasn't sure given the setup.

shakuzen replied:
I tried to do a `pTML` (publishToMavenLocal) which is what I do to test changes in snapshots locally, but it failed because the signing properties aren't configured. We want contributors to be able to publish local snapshots without needing to configure signing keys. That said, we do want to ensure signing is done for remote publishing. I'm not sure the best way to go about that. I added
```
if (hasProperty('SIGNING_KEY') && hasProperty('SIGNING_PASSWORD')) {
```
but that makes it possible for remote publishing to be attempted without signing artifacts, which I think we want to avoid - it will fail Maven Central's requirements, for example, and is just good practice if we easily can.

jonatan-ivanov replied:
Oops, I messed this up. This signs snapshots too as well as local publications.
I fixed it this way:
```groovy
signing {
            required = hasProperty('SIGNING_KEY')
}
```

This tells the signing plugin that signature is required when the key is present otherwise the task will be skipped so if you have the key it will sign local builds, if you don't it won't. I did not include the password since the gpg key can be passwordless.

shakuzen replied:
That fixes the issue of contributors being able to publish a snapshot to their local maven repository. We should probably still think about checking that signing is configured if we're publishing a release to a remote repository.

jonatan-ivanov commented:
@shakuzen  Changes since the last review:
- `closeAndReleaseMavenCentralStagingRepository` task (https://github.com/gradle-nexus/publish-plugin)
- snapshot is published the same way as everything else (maven-publish instead of artifactory)
- `publishAllPublicationToMavenCentralRepository` -> `publishNebulaPublicationToMavenCentralRepository`
- optional signature: `publishToMavenLocal` works without a key
- `deploy.sh` matches on milestone now instead of rc (I tested it, it works)

Also, we can decide if we want to release manually or automatically: `closeMavenCentralStagingRepository` vs. `closeAndReleaseMavenCentralStagingRepository`. Right now, the latter is used.
